src/Encoding.chpl

```diff
@@ -0,0 +1,248 @@
+    var uintVal = (val << 1):uint ^ (val >> 31):uint;
+    unsignedVarintAppend(uintVal, ch);
+  }
+
+  proc sint32Consume(ch: readingChannel): int(32) throws {
+    const (val, len) = unsignedVarintConsume(ch);
+    return (val >> 1):int(32) ^ (val):int(32) << 31 >> 31;
+  }
+
+  proc bytesAppend(val: bytes, ch: writingChannel) throws {
+    const wireType = lengthDelimited;
+    unsignedVarintAppend((val.size):uint, ch);
+    ch.write(val);
+  }
+
+  proc bytesConsume(ch: readingChannel): bytes throws {
+    const (byteLen, len) = unsignedVarintConsume(ch);
+    var s:bytes;
+    ch.readbytes(s, byteLen:int);
+    return s;
+  }
+
+  proc stringAppend(val: string, ch: writingChannel) throws {
+    bytesAppend(val.encode(), ch);
+  }
+
+  proc stringConsume(ch: readingChannel): string throws {
+    return bytesConsume(ch).decode();
+  }
+
+
+  proc fixed32Append(val: uint(32), ch: writingChannel) throws {
+    const wireType = fixed32Type;
+    ch.write(val);
+  }
+
+  proc fixed32Consume(ch: readingChannel): uint(32) throws {
+    var val: uint(32);
+    ch.read(val);
+    return val;
+  }
+
+  proc fixed64Append(val: uint(64), ch: writingChannel) throws {
+    const wireType = fixed64Type;
+    ch.write(val);
+  }
+
+  proc fixed64Consume(ch: readingChannel): uint(64) throws {
+    var val: uint(64);
+    ch.read(val);
+    return val;
+  }
+
+  proc floatAppend(val: real(32), ch: writingChannel) throws {
+    var a = val;
+    var b: uint(32);
+    c_memcpy(c_ptrTo(b), c_ptrTo(a), c_sizeof(b.type));
+    fixed32Append(b, ch);
+  }
+  
+  proc floatConsume(ch: readingChannel): real(32) throws {
+    var a = fixed32Consume(ch);
+    var b: real(32);
+    c_memcpy(c_ptrTo(b), c_ptrTo(a), c_sizeof(b.type));
+    return b;
+  }
+
+  proc doubleAppend(val: real(64), ch: writingChannel) throws {
+    var a = val;
+    var b: uint(64);
+    c_memcpy(c_ptrTo(b), c_ptrTo(a), c_sizeof(b.type));
+    fixed64Append(b, ch);
+  }
+  
+  proc doubleConsume(ch: readingChannel): real(64) throws {
+    var a = fixed64Consume(ch);
+    var b: real(64);
+    c_memcpy(c_ptrTo(b), c_ptrTo(a), c_sizeof(b.type));
+    return b;
+  }
+
+  proc sfixed64Append(val: int(64), ch: writingChannel) throws {
+    var a = val;
+    var b: uint(64);
+    c_memcpy(c_ptrTo(b), c_ptrTo(a), c_sizeof(b.type));
+    fixed64Append(b, ch);
+  }
+
+  proc sfixed64Consume(ch: readingChannel): int(64) throws {
+    var a = fixed64Consume(ch);
+    var b: int(64);
+    c_memcpy(c_ptrTo(b), c_ptrTo(a), c_sizeof(b.type));
+    return b;
+  }
+
+  proc sfixed32Append(val: int(32), ch: writingChannel) throws {
+    var a = val;
+    var b: uint(32);
+    c_memcpy(c_ptrTo(b), c_ptrTo(a), c_sizeof(b.type));
+    fixed32Append(b, ch);
+  }
+
+  proc sfixed32Consume(ch: readingChannel): int(32) throws {
+    var a = fixed32Consume(ch);
+    var b: int(32);
+    c_memcpy(c_ptrTo(b), c_ptrTo(a), c_sizeof(b.type));
+    return b;
+  }
+
+  proc writeToOutputFileHelper(ref message, ch) {
```

Comment:
Shouldn't this function be marked `throws` ? As well as the read version and the code calling it generated by the protobuf compiler?